src/overlays/actors/ovl_Obj_Mine/z_obj_mine.c

```diff
@@ -24,131 +56,1178 @@ ActorInit Obj_Mine_InitVars = {
+void ObjMine_Water_SetWorld(ObjMine* this) {
+    s32 pad; // Could be recast to thisx as in ObjMine_Path_Move
+    ObjMineWaterChain* waterChain = &this->chain.water;
+    s32 linkCount = OBJMINE_GET_LINK_COUNT(&this->actor);
+
+    if (linkCount == 0) {
+        this->actor.world.pos.y = this->actor.home.pos.y + LINK_SIZE + ATTACH_OFFSET;
+    } else {
+        ObjMineWaterLink* lastLink = &waterChain->links[linkCount - 1];
+        Vec3f mineOffset;
+
+        Math_Vec3f_ScaleAndStore(&lastLink->basis.y, ATTACH_OFFSET, &mineOffset);
+        Math_Vec3f_Sum(&lastLink->pos, &mineOffset, &this->actor.world.pos);
+    }
+}
+
+void ObjMine_Water_WallCheck(ObjMine* this, PlayState* play) {
+    s32 pad; // Could be recast to thisx as in ObjMine_Path_Move
+    ObjMineWaterChain* waterChain = &this->chain.water;
+
+    waterChain->touchWall = false;
+    if (waterChain->wallCheckDistSq > -1e-6f) {
+        //  Checks for walls if mine is sufficiently far from home.
+        if (waterChain->wallCheckDistSq <= Math3D_XZDistanceSquared(this->actor.home.pos.x, this->actor.home.pos.z,
+                                                                    this->actor.world.pos.x, this->actor.world.pos.z)) {
+            Vec3f centerPos;
+            Vec3f offsetPos;
+            Vec3f dummyResult;
+            Vec3f xzFromHome;
+            Vec3f xzDirFromHome;
+            CollisionPoly* dummyPoly;
+            s32 dummyBgId;
+            f32 dummyNorm;
+            f32 dummyInvNorm;
+
+            xzFromHome.x = this->actor.world.pos.x - this->actor.home.pos.x;
+            xzFromHome.y = 0.0f;
+            xzFromHome.z = this->actor.world.pos.z - this->actor.home.pos.z;
+
+            if (ObjMine_GetUnitVec3fNorm(&xzFromHome, &xzDirFromHome, &dummyNorm, &dummyInvNorm)) {
+
+                offsetPos.x = this->actor.world.pos.x + (xzDirFromHome.x * WATER_RADIUS);
+                offsetPos.y = this->actor.world.pos.y;
+                offsetPos.z = this->actor.world.pos.z + (xzDirFromHome.z * WATER_RADIUS);
+
+                centerPos.x = this->actor.home.pos.x;
+                centerPos.y = this->actor.world.pos.y;
+                centerPos.z = this->actor.home.pos.z;
+
+                if (BgCheck_EntityLineTest1(&play->colCtx, &centerPos, &offsetPos, &dummyResult, &dummyPoly, true,
+                                            false, false, true, &dummyBgId)) {
+                    waterChain->touchWall = true;
+                    waterChain->wallEject.x = xzDirFromHome.x * -0.2f;
+                    waterChain->wallEject.z = xzDirFromHome.z * -0.2f;
+                }
+            }
+        }
+    }
+}
+
+#ifdef NON_MATCHING
+// Probably equivalent, but nightmare loop unrolls. The tension section also has weird comparisons. It should be
+// equivalent to what's written, but it seems the actual code comparisons were implemented differently. ASM suggests
+// multiple index variables were used at once.
+void ObjMine_Water_ApplyForces(ObjMine* this) {
```

Comment:
Could you post a decomp.me scratch for this function?

Reply:
Added to the PR and to the source.